micro-rdk/src/common/config_monitor.rs

```diff
@@ -0,0 +1,56 @@
+use super::app_client::{ AppClient, AppClientError, PeriodicAppClientTask};
+use futures_lite::Future;
+use std::pin::Pin;
+use std::time::Duration;
+use crate::proto::app::v1::ConfigResponse;
+
+pub struct ConfigMonitor<'a> {
+    restart_hook: Option<Box<dyn FnOnce() + 'a>>,
+    curr_config: ConfigResponse, //config for robot gotten from last robot startup, aka inputted from entry
+}
+
+impl<'a> ConfigMonitor<'a> {
+    pub fn new(restart_hook: impl FnOnce() + 'a, curr_config: ConfigResponse) -> Self {
+        Self {
+            restart_hook: Some(Box::new(restart_hook)),
+            curr_config: curr_config,
+        }
+    }
+
+    fn restart(&mut self) -> ! {
+        log::warn!("Config change detected - restarting or terminating now...");
+        (self.restart_hook.take().unwrap())();
+        unreachable!();
+    }
+}
+
+impl<'a> PeriodicAppClientTask for ConfigMonitor<'a> {
+    fn name(&self) -> &str {
+        "ConfigMonitor"
+    }
+
+    fn get_default_period(&self) -> Duration {
+        Duration::from_secs(10)
+    }
+
+    fn invoke<'c, 'b: 'c>(
+        &'b mut self,
+        app_client: &'c AppClient,
+    ) -> Pin<Box<dyn Future<Output = Result<Option<Duration>, AppClientError>> + 'c>> {
+        Box::pin(async move {
+            let app_config = get_app_config(app_client).await.unwrap();
+            match self.curr_config == app_config{
+                true => self.restart(),
+                false => Ok(Some(self.get_default_period())),
+            }
+        })
+    }
+
+}
+
+async fn get_app_config(client : &AppClient) -> Result<ConfigResponse, AppClientError>{
+    let (new_config, _cfg_received_datetime) = 
+            client.get_config(None).await.unwrap();
+    return Ok(*new_config);
+}
```

Comment:
I'm not sure this function is pulling its weight. Is there a reason the body can't just be inlined into `invoke` above?

Reply:
I don't think you need to make `invoke` an async function: it already "is" one in the sense that it returns a Future, but it can't be declared as one due to limitations in rust. However, note that on line 40 you are inside an `async` block, and as far as I know, inside there you can definitely use `await` (it already does).